internal/core/services/autoscaler/policies/roomoccupancy/policy.go

```diff
@@ -0,0 +1,88 @@
+// MIT License
+//
+// Copyright (c) 2021 TFG Co
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package roomoccupancy
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+
+	"github.com/topfreegames/maestro/internal/core/entities"
+	"github.com/topfreegames/maestro/internal/core/entities/game_room"
+	"github.com/topfreegames/maestro/internal/core/ports"
+	autoscalerPorts "github.com/topfreegames/maestro/internal/core/ports/autoscaler"
+	"github.com/topfreegames/maestro/internal/core/services/autoscaler/policies"
+)
+
+const (
+	// ReadyTargetKey is the key to ready target in the CurrentState map.
+	ReadyTargetKey = "RoomsOccupancyReadyTarget"
+	// OccupiedRoomsKey is the key to occupied rooms in the CurrentState map.
+	OccupiedRoomsKey = "RoomsOccupancyOccupiedRooms"
+)
+
+// Policy holds the requirements to build the current state of
+// the scheduler that should be considered to calculate the desired number of rooms in the room occupancy policy.
+type Policy struct {
+	roomStorage ports.RoomStorage
+}
+
+var _ autoscalerPorts.Policy = new(Policy)
+
+// NewPolicy create a new room occupancy autoscaling policy.
+func NewPolicy(roomStorage ports.RoomStorage) *Policy {
+	return &Policy{
+		roomStorage: roomStorage,
+	}
+}
+
+// CurrentStateBuilder fill the fields that should be considered during the autoscaling policy.
+func (p *Policy) CurrentStateBuilder(ctx context.Context, scheduler *entities.Scheduler) (policies.CurrentState, error) {
+	occupiedRoomsAmount, err := p.roomStorage.GetRoomCountByStatus(ctx, scheduler.Name, game_room.GameStatusOccupied)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching occupied game rooms amount: %w", err)
+	}
+
+	readyTarget := scheduler.Autoscaling.Policy.Parameters.RoomOccupancy.ReadyTarget
+
+	currentState := policies.CurrentState{
+		ReadyTargetKey:   readyTarget,
```

Comment:
Ready target isn't really part of the current state, right? it is an autoscaling policy parameter. Why configure it here since we already have access to it in `Autoscaling` struct configuration? 

What about updating the CalculateDesiredNumberOfRooms method to receive the current state and the **policy parameters** since all policies will have to use its parameter to calculate the desired number of rooms:

`CalculateDesiredNumberOfRooms(currentState policies.CurrentState, policyParameters PolicyParameters) (int, error)`